[all] Add namespaces in the factory and a python-like import mechanism

damienmarchal opened the pull request:
This PR aims to handle in a more "contained" way the components implemented in different plugins to avoid having everything mixed into the ObjectFactory. 

The PR is composed of two aspects.

1) **Registering components by "plugin.classname" instead of "classname" in the ObjectFactory**

The factory currently access component by their classname only causing collision when two component have the same name/template. To avoid this, the PR add the plugin name as a prefix for the Factory'key. As an exemple, a component created with *<OglModel>* in the past is now referenced as *<SofaOpenglVisual.OglModel>*.

Backward compatbility is possible by adding  into the factory registration an alias mapping between 
new names and old name;
Eg:
```
"OglModel" => "SofaOpenglVisual.OglModel"
```

2) **Importing plugin and component *à la* python**
A new component called *FromComponent* is added to handle the loading of plugin and the import of object. 

Examples of use:
```
	 # To make only <OglModel> accessible in the scene
	<From plugin="SofaOpenglVisual" import="OglModel"/>
	<OglModel/>

	# To add "custom" name in a scene 
	<From plugin="SofaOpenglVisual" import="OglModel" as="RenderingModel"/>
	<RenderingModel/>

	# To be a super lazy man and import everything
	<From plugin="SofaOpenglVisual" import="*"/>
```

I didn't implemented multiple object imports so currently it is necessary to do 
```
       <From plugin="SofaOpenglVisual" import="OglModel"/>
       <From plugin="SofaOpenglVisual" import="GlslShader"/>
```
But we could as in python with:
```
       <From plugin="SofaOpenglVisual" import="OglModel, GlslShader"/>
```


______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

damienmarchal committed:
[SofaBaseLinearSolver] Remove the IsRequired flag in CGLinearSolver

PR 2419 restores the correct behavior of CGLinearSolver regarding the call-super-init
https://github.com/sofa-framework/sofa/pull/2419

But as some of its data are tagged as Required, this rise en error message when the value
is not set by the user.

I'm not sure it is wise to use the required flag when data provides meaningful and usable default values.
So I recommand removing it.

damienmarchal committed:
[SofaKernel] Add a namespace like behavior to factory & "à la" python import

This PR aims at handling in a proper way the components implemented in different plugings.
The PR is composed of two aspects.

1) Registering components by "plugin.classname" in the Factory
The factory currently access component by their classname only causing
collision when two component have the same name/template.
To avoid this, the PR add the plugin name as a prefix for the Factory'key.
So as an exemple, a component created with <OglModel> in the past is now referenced
as <SofaOpenglVisual.OglModel>.

Backward compatbility is possible by hard coding into the factory registration the adding of an alias from
new names to old name;
Eg adding an alias saying:
   "OglModel" => "SofaOpenglVisual.OglModel"

This makes the change 100% compatible with existing sofa scenes.

2) Importing plugin and component in a clear way (à la python)
The PR add a new component (FromComponent) that import one or several components.

Examples of use:
	 # Only <OglModel> is accessible in the scene
	<From plugin="SofaOpenglVisual" import="OglModel"/>
	<OglModel/>

	# I like to have my own "custom" name in my scene because I ate "official ones"
	<From plugin="SofaOpenglVisual" import="OglModel" as="VisualModel"/>
	<From plugin="SofaOpenglVisual" import="OglModel" as="RenderingModel"/>
	<VisualModel/>
	<RenderingModel/>


	# I'm a super lazy man so I import everything
	<From plugin="SofaOpenglVisual" import="*"/>

Signed-off-by: Damien Marchal <damien.marchal@univ-lille1.fr>

damienmarchal committed:
[SofaKernel] Add a mecanisme to handle alias deprecation.

A list is holding if an alias should emit a warning or not.
Allowed alias are the one defined in the scene.

damienmarchal committed:
[SofaKernel] Fix error message name in FromComponent

damienmarchal committed:
[SofaKernel] Use the newly introduced target.name registration for the RequiredPlugin component;

The RequiredPlugin is now by default mapped to its full name:
SofaBaseUtils.RequierdPlugin

The alias name "RequiredPlugin" is not generating a warning.

damienmarchal committed:
[SofaGraphComponent] Add the new alias deprecation system into SceneCheckUsingAlias.cpp

damienmarchal committed:
[SofaBaseutils_test] Add tests for the "From" component.

damienmarchal commented:
I all,

I have improved the whole mecanism so now it is off by default and we can activate it selectively (progressively) 
per component during their registration in the factory.

I also added a mecanism to handle in a smooth way the deprecation of aliases as well as a update the SceneCheckUsingAlias checker. 

I don't see a lot of drawback for all this work, while I see a lot of benefits, so feedback welcome.

damienmarchal committed:
[SofabaseUtils_test] Add missing From_test.cpp file.

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-add-import-from

damienmarchal committed:
[SofaCore] Fix link error on windows.

alxbilger commented on SofaKernel/modules/SofaBaseUtils/src/SofaBaseUtils/FromComponent.cpp:
```diff
@@ -0,0 +1,136 @@
+
+    // First load the plugin if it is not yet done
+    type::vector< std::string > failed;
+    std::ostringstream errmsg;
```
This variable can be moved into the next condition

alxbilger commented on SofaKernel/modules/SofaBaseUtils/src/SofaBaseUtils/FromComponent.cpp:
```diff
@@ -0,0 +1,136 @@
+        ObjectFactory::getInstance()->getEntriesFromTarget(entries, std::string(plugin));
+        for(auto& entry : entries)
+        {
+            string fullname = std::string(plugin)+"."+entry->className;
```
```suggestion
            const std::string fullname = std::string(plugin)+"."+entry->className;
```

alxbilger commented on SofaKernel/modules/SofaBaseUtils/src/SofaBaseUtils/FromComponent.cpp:
```diff
@@ -0,0 +1,136 @@
+    std::string fullname = tmp.str();
+    if(!ObjectFactory::getInstance()->hasCreator(fullname))
+    {
+        std::stringstream tmp;
```
`tmp` is already declared and I recommend a more explicit name. `errorMessage` for example

alxbilger commented on SofaKernel/modules/SofaBaseUtils/src/SofaBaseUtils/FromComponent.cpp:
```diff
@@ -0,0 +1,136 @@
+    d_componentState.setValue(ComponentState::Valid) ;
+}
+
+int FromComponentClass = RegisterObject("This object create an alias to a component name to make the scene more readable. ")
```
```suggestion
int FromComponentClass = RegisterObject("This object creates an alias to a component name to make the scene more readable. ")
```

alxbilger commented on modules/SofaGraphComponent/src/SofaGraphComponent/SceneCheckUsingAlias.cpp:
```diff
@@ -39,9 +39,14 @@ SceneCheckUsingAlias::SceneCheckUsingAlias()
         if ( typeNameInScene != o->getClassName() )
         {
-            this->m_componentsCreatedUsingAlias[o->getClassName()].push_back(typeNameInScene);
+            // Warnings are eimtted only for the alias that are flagged as deprecated;
```
```suggestion
            // Warnings are emitted only for the alias that are flagged as deprecated
```

alxbilger commented on modules/SofaGraphComponent/src/SofaGraphComponent/SceneCheckUsingAlias.cpp:
```diff
@@ -39,9 +39,14 @@ SceneCheckUsingAlias::SceneCheckUsingAlias()
         {
-            this->m_componentsCreatedUsingAlias[o->getClassName()].push_back(typeNameInScene);
+            // Warnings are eimtted only for the alias that are flagged as deprecated;
+            auto entry = ObjectFactory::getInstance()->getEntry(o->getClassName());
```
```suggestion
            const auto& entry = ObjectFactory::getInstance()->getEntry(o->getClassName());
```

alxbilger commented on SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.h:
```diff
@@ -91,12 +91,16 @@ class SOFA_CORE_API ObjectFactory
         CreatorMap creatorMap;
         std::map<std::string, std::vector<std::string>> m_dataAlias ;
+
+        bool SOFA_CORE_API isADeprecatedAlias(const std::string& name);
```
```suggestion
        bool SOFA_CORE_API isADeprecatedAlias const (const std::string& name);
```

alxbilger commented on SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.cpp:
```diff
@@ -29,6 +29,11 @@
 namespace sofa::core
 {
 
+bool ObjectFactory::ClassEntry::isADeprecatedAlias(const std::string& name)
```
```suggestion
bool ObjectFactory::ClassEntry::isADeprecatedAlias(const std::string& name) const
```

alxbilger commented on SofaKernel/modules/SofaBaseUtils/src/SofaBaseUtils/FromComponent.cpp:
```diff
@@ -0,0 +1,136 @@
+                           "To remove this error message you need to add a targetcomponent attribute pointing to a valid component's ClassName.";
+        return ;
+    }
+    string starget(plugin) ;
```
I think it is not used
```suggestion
```

alxbilger commented on SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.h:
```diff
@@ -91,12 +91,16 @@ class SOFA_CORE_API ObjectFactory
         std::string authors;
         std::string license;
         std::string defaultTemplate;
+        std::string compilation_target;
```
```suggestion
        std::string compilationTarget;
```
To follow guidelines and keep consistency with other class member names.

alxbilger reviewed (changes requested):
I have some questions:

The `RequiredPlugin` and `FromComponent` components are the only ones to use `addTargetName`, which is the only function to set `compilation_target`. Does it mean that all the other components don't register their target name? Should all the components use `addTargetName` when they register to the factory?

damienmarchal commented:
Hello @alxbilger,

Thank for the careful review, I will integrate that if the feature is considered worth for inclusion. 

About the "addTargetName" question. 

The initial version proposed was using automatic registration of the target name and I can revert back to this one. 
But I proposed a second, explicit, approach for two reasons:
- the automatic mecanism force the deployement of the new approach for all the sofa code components at once... which I was not sure we agreed at this point. 
- the existing mecanism to define the target in ObjectFactory.h is buggy because two identical ObjectFactory::register<XX> compiled in different compilation units produces different getTarget() functions. There was an agreement we should remove it, so this PR was a good opporunity to trying such a thing. 

Hope this clarify things.

EDIT:
I just found a funny bug related to the automatic registration. Because it rely on the use of a backward macro expansion. There is currently more than 20 components that does not have their target properly set.
Eg: DataDisplay.cpp the code is like that:
```cpp
#include <sofa/core/ObjectFactory.h>
#include <sofa/defaulttype/VecTypes.h>

#include <SofaOpenglVisual/config.h>
#include <SofaOpenglVisual/DataDisplay.h>
#include <sofa/core/visual/VisualParams.h>

namespace sofa::component::visual::model
{
     int DataDisplayClass = core::RegisterObject("Rendering of meshes colored by data")
        .add< DataDisplay >();
}
```

The problem is that the ```SOFA_TARGET``` macro is used in ```ObjetFactory``` but is actually initialized in ```<SofaOpenglVisual/config.h>```
To fix the code it is needed to move the inclusion of the config.h before the one of ObjectFactory so that RegisterObject behave as expected. 
```cpp
#include <SofaOpenglVisual/config.h>
#include <sofa/core/ObjectFactory.h>
#include <sofa/defaulttype/VecTypes.h>

#include <SofaOpenglVisual/DataDisplay.h>
#include <sofa/core/visual/VisualParams.h>
```

Given all that, I'm wonder if going for things like  
```cpp
     int DataDisplayClass = core::RegisterObject("Rendering of meshes colored by data")
        .add< DataDisplay >();
        .addTargetName("SofaOpenglVisual")
```        
or even:
```cpp
     int DataDisplayClass = core::RegisterObject()
         .setClassName("DataDisplay"), 
         .setTargetName("SofaOpenglVisual"),           
         .setDocString("Rendering of meshes colored by data")
         .addCreator<DataDisplay>();
```        
Would not make things much more clear.

damienmarchal committed:
Update SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.h

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

damienmarchal committed:
Update SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.cpp

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

damienmarchal committed:
Update modules/SofaGraphComponent/src/SofaGraphComponent/SceneCheckUsingAlias.cpp

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

damienmarchal committed:
Update modules/SofaGraphComponent/src/SofaGraphComponent/SceneCheckUsingAlias.cpp

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

damienmarchal committed:
Update SofaKernel/modules/SofaBaseUtils/src/SofaBaseUtils/FromComponent.cpp

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

damienmarchal committed:
Update SofaKernel/modules/SofaBaseUtils/src/SofaBaseUtils/FromComponent.cpp

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

damienmarchal committed:
Merge remote-tracking branch 'cristal/pr-add-import-from' into pr-add-import-from

damienmarchal committed:
[SofaKernel] Integrate suggestion made by @alxbilger in PR #2512

damienmarchal committed:
[SofaKernel] Add a warning if component is registered without a compilation target.

damienmarchal commented:
The failing test is actually showing there was a bug before :)

damienmarchal committed:
[SOfaGeneralLoader] Fix invalid target in GIDMeshLoader.cpp

The definition of SOFA_TARGET definition needs to be done before
the #include of <ObjectFactory.h>

damienmarchal committed:
Merge remote-tracking branch 'cristal/pr-add-import-from' into pr-add-import-from

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-add-import-from

fredroy commented on SofaKernel/modules/SofaBaseUtils/src/SofaBaseUtils/FromComponent.h:
```diff
@@ -0,0 +1,68 @@
+
+    Data<std::string>   d_plugin                ; ///< The component class for which to create an alias.
+    Data<std::string>   d_import_old_name       ; ///< The object's name to import
+    Data<std::string>   d_as_new_name           ; ///< The new alias of the component.
```
```
    Data<std::string>   d_importOldName       ; ///< The object's name to import
    Data<std::string>   d_asNewName           ; ///< The new alias of the component.
```
lowerCamelCase guideline

I did not put a suggestion as it will induce modifications further in the code

damienmarchal replied:
ARgg, too much python pep8 recently 
thanks for feedback.

fredroy commented:
[ci-build][with-all-tests]

damienmarchal commented:
Replace ```<From ...>```
By: ```<Import from="blabla" component="VisualModel" as="Toto"/>```

And multiple load with comma separation. 
"toto tata titi ti ti" 
" 0 1 2 3 4" 

=====================
Compatibility with 
SOFA NG:
```<From plugin="Sofa.Component.ODESolver" import="*"/>``` => OK ?  Sofa.Component.ODESolver.Backward
```<From plugin="Sofa.Component" import="*"/>``` => OK ?   

In an xml Sofa scène: Sofa.Component.ODESolver.Backward
In a python sofa scene: Sofa.Component.ODESolver.Backward 
In the c++ namespace: sofa::component::odesolver::Backward
In the c++ include: sofa/component/odesolver/Backward
...
All that is so nicely consistent.

damienmarchal commented:
Several alternatives choices for names:
```xml
<From plugin="SofaVisual" import="OglModel" as="VisualModel"/>
<FromPlugin name="SofaVisual" import="OglModel" as="VisualModel"/>

<Plugin name="SofaVisual" import="OglModel" as="VisualModel"/>
<Plugin import="OglModel" as="VisualModel" from="SofaVisual" />

<Import from="SofaVisual" component="OglModel" as="VisualModel"/>      
<Import from="SofaVisual" components="*"/>

<Import fromPlugin="SofaVisual" components="OglModel OglShader"/>

```

damienmarchal committed:
Remove passager clandestin.

damienmarchal committed:
Add missing files.

damienmarchal committed:
Rename From as Import.

damienmarchal commented:
So I finally used the following name
Valid cases
```xml
    <Import fromPlugin="SofaOpenglVisual" component="OglModel"/>
    <Import fromPlugin="SofaOpenglVisual" component="OglModel" as="MyVisualModel"/>
    <Import fromPlugin="SofaOpenglVisual" components="OglModel,OglShader"/>
    <Import fromPlugin="SofaOpenglVisual" components="OglModel, OglShader"/>
    <Import fromPlugin="SofaOpenglVisual" components="*"/>
```

Invalid cases
``` 
    <Import fromPlugin="SofaOpenglVisual"/>                                                                                            
    <Import fromPlugin="SofaOpenglVisual" component="OglModel" components="OglShader"/>
    <Import fromPlugin="SofaOpenglVisual" components="OglModel" as="VisualModel"/>
    <Import fromPlugin="SofaOpenglVisual" components="OglModel, OglShader" as="Invalid1, Invalid2"/>  // Maybe this one ?
    <Import fromPlugin="SofaOpenglVisual" components="*" as="Test"/>
```

damienmarchal committed:
Change docstring in ImportComponent

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-add-import-from

damienmarchal committed:
Rename From => Import

damienmarchal committed:
Minor cleaning in ImportComponent.

damienmarchal commented:
Hi @hugtalbot what does it means "tested first" ?

alxbilger commented:
```xml
<RequiredPlugin name="MultiThreading"/>
<BruteForceBroadPhase/>
<BVHNarrowPhase/>
```
loads the single threaded versions


```xml
<RequiredPlugin name="MultiThreading"/>
<MultiThreading.ParallelBruteForceBroadPhase/>
<MultiThreading.BVHNarrowPhase/>
```
loads the multithreaded versions

```xml
<Import fromPlugin="MultiThreading" components="ParallelBruteForceBroadPhase" as="BruteForceBroadPhase"/>
<BruteForceBroadPhase/>
<BVHNarrowPhase/>
```
loads the ParallelBruteForceBroadPhase component, not the BruteForceBroadPhase component.

alxbilger commented:
@damienmarchal:

- BVHNarrowPhase is in SofaBaseCollision
- ParallelBVHNarrowPhase is in MultiThreading and inherits from BVHNarrowPhase

Can I add ParallelBVHNarrowPhase as BVHNarrowPhase in the ObjectFactory? We would have SofaBaseCollision.BVHNarrowPhase and MultiThreading.MultiThreading. Does it make sense to do that?

damienmarchal commented:
Hi @alxbilger,

Thanks for the tests, which if I understand correctly are working as expected. 

When doing
```xml
<Import fromPlugin="MultiThreading" components="ParallelBVHNarrowPhase" as="BVHNarrowPhase"/>
```
The alias BVHNarrowPhase that is initially pointing to SofaBaseCollision.BVHNarrowPhase should be replaced with one with the same named BVHNarrowPhase but pointing to Multithread.BVHNarrowPhase.

guparan commented on SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.cpp:
```diff
@@ -580,10 +596,21 @@ RegisterObject::operator int()
+        reg.deprecatedAliases = entry.deprecatedAliases;
+
+        if(reg.compilationTarget.empty())
+            msg_warning("ObjectFactory") << "No compilationTarget for object " << entry.className;
```
Add a "fixme" message

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-add-import-from

damienmarchal committed:
[SofaKernel] Improve error message.

guparan reviewed (commented):
Actually I don't approve, some tests are failing on CI :(

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-add-import-from

damienmarchal commented:
Hi all,

Here are the result why BarycentricMapping<> is failing in the BarycentricMappingTruss example.

Summary on how object are registered in the factory in sofa/master: 
1) BarycentricMapping<Vec3d,Vec3d> is added into the factory using "BarycentricMapping" as entry name
Then 
2) BarycentricMapping<Vec3d, Rigid3d> is aded into the factory using "BarycentricMapping" as entry name (so this actually adds a new template to the existing one)


With the PR this is what happens:
1) BarycentricMapping<Vec3d,Vec3d> is added into the factory using "SofaBaseMechanics.BarycentricMapping" as entry name. 
2) Then, for compatibility, it adds an alias from BarycentricMapping to SofaBaseMechanics.BarycentricMapping. 
3) Then BarycentricMapping<Vec3d, Rigid3d> is added into the factory using "SofaMiscMapping.BarycentricMapping" as entry name. This is actually correct (as SofaMiscMapping is the target where this mapping lies in).

But the alias is not updated to point both to SofaBaseMechanics.BarycentricMapping and SofaMIscMapping.BarycentricMapping...which, actually seems pretty logic...one name - one object. 

Maybe we can change the alias mechanism so we allow an alias to point to multiple different object...which I'm not very convinced this will not get into troubles.

damienmarchal committed:
WIP

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-add-import-from

damienmarchal committed:
[SofaCore] Implement the alias properly.

damienmarchal committed:
[SofaSimulatinCommon] Cosmetic cleaning in ElementNameHelper::resolveName

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-add-import-from

# Conflicts:
#	Component/IO/Mesh/src/sofa/component/io/mesh/GIDMeshLoader.cpp
#	SofaKernel/modules/SofaBaseUtils/CMakeLists.txt

damienmarchal committed:
Remove getTarget.

damienmarchal committed:
[all] Fix missing #include config.h (so the factory's getTarget) is not correct)

damienmarchal committed:
[SofaSimulation] Remove DAGNodeMultiMappingElement

I have no idea what this code is supposed to do.
It does not seems related to something that is in SimulationGraph.
Removing it does not change anything.

damienmarchal committed:
WIP on the factory.

damienmarchal committed:
[SofaCore] Add missing ObjectFactory_test.cpp
